miden-lib/asm/sat/prologue.masm

```diff
@@ -136,6 +171,65 @@ end
 # CONSUMED NOTES DATA
 # =================================================================================================
 
+#! Authenticate the consumed note data provided via the advice provider is consistent with the
+#! the chain history.  This is achieved by:
+#! - authenticating the MMR leaf associated with the block the note was created in.
+#! - authenticating the note root associated with the block the note was created in.
+#! - authenticating the note and its metadata in the note Merkle tree from the block the note was
+#!   created in.
+#!
+#! Stack: [AUTH_DIGEST]
+#! Advice Stack: [leaf_pos, SUB_HASH, NOTE_ROOT, note_index]
+#! Output: []
+#!
+#! - AUTH_DIGEST is the digest of the consumed note data computed as hash(NOTE_HASH, NOTE_METADATA)
+#! - leaf_pos is the position of the leaf in the MMR associated with the block the note was created
+#!   in. This is equivalent to the block number.
+#! - SUB_HASH is the sub hash of the block the note was created in.
+#! - NOTE_ROOT is the note root of the block the note was created in.
+#! - note_index is the index of the note in the note Merkle tree.
+proc.authenticate_note.2
+    # load data required for MMR get operation
+    exec.layout::get_chain_mmr_ptr adv_push.1
+    # => [leaf_pos, chain_mmr_ptr, AUTH_DIGEST]
+
+    # get the chain MMR leaf associated with the block the note was created in
+    exec.mmr::get
+    # => [MMR_LEAF, AUTH_DIGEST]
+
+    # prepare the stack to read the sub hash and note root from the advice provider
+    locaddr.0 padw padw padw
+    # => [PAD, PAD, PAD, mem_ptr, MMR_LEAF, AUTH_DIGEST]
+
+    # read the core hash and note root from the advice provider
+    adv_pipe
+    # => [PERM, PERM, PERM, mem_ptr', MMR_LEAF, AUTH_DIGEST]
+
+    # extract the digest and assert it matches MMR_LEAF
+    dropw swapw dropw movup.4 drop assert_eqw
+    # => [AUTH_DIGEST]
+
+    # load the note root from memory
+    padw loc_loadw.1 swapw
+    # => [AUTH_DIGEST, NOTE_ROOT]
+
+    # load the index of the note
+    adv_push.1 movdn.4
+    # => [AUTH_DIGEST, note_index, NOTE_ROOT]
+
+    # get the depth of the note tree
+    exec.constants::get_note_tree_depth sub.1 movdn.4
+    # => [AUTH_DIGEST, depth, note_index, NOTE_ROOT]
```

Comment:
I'm wondering: should we assume here that the depth goes to `AUTH_DIGEST`? This way, we can have $2^{20}$ notes in the tree (rather than $2^{19}$) and also don't need to do `sub.1` here.

Reply:
Yep this sounds reasonable.